### PR TITLE
[MRG] Use exact types for Py_BuildValue.

### PR DESCRIPTION
**Pull request by QuLogic**
IIRC, most ABI upcast values passed to vararg functions anyway, but there might be some other ABIs that require the exact correct type.

### CODE REVIEW

**Comment by story645** [comment]
Stupid questions cause I wanna wrap my head around this... What's the reference you're using to get specific types? And how come the new specific types don't have to be the same size as (and sometimes bigger than) the old types?

**Comment by QuLogic** [comment]
I get the types from the [FreeType API](https://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html).

IIRC, C ABI promotes all scalar values into the largest compatible type when passed to vararg functions like `Py_BuildValue`, so there's generally always enough space to treat it as a larger type even when it didn't start as one. I know for certain that this is true of `float`, which is promoted to `double`, on x86*, but I'm not exactly sure about integral types. And the ABI may be different on other architectures.

File: src/ft2font.h [hunk 1]
```diff
@@ -18,8 +18,8 @@ extern "C" {
 /*
  By definition, FT_FIXED as 2 16bit values stored in a single long.
  */
-#define FIXED_MAJOR(val) (long)((val & 0xffff000) >> 16)
-#define FIXED_MINOR(val) (long)(val & 0xffff)
+#define FIXED_MAJOR(val) (unsigned short)((val & 0xffff000) >> 16)
```
Comment (NelleV) [inline src/ft2font.h]: Isn't that supposed to be a signed long?

Comment (QuLogic) [inline src/ft2font.h]: Not exactly, but there appears to maybe be a deeper existing bug here. I'm not sure if I'm just reading it wrong, but will need some time to investigate.

Comment (story645) [inline src/ft2font.h]: What are `FIXED_MAJOR` and `FIXED_MINOR`?

Comment (QuLogic) [inline src/ft2font.h]: They extract the two 16-bit values. However, there's a bug in `FIXED_MAJOR`; it masks the input as if it were 16.12 (with 4 undefined bits) yet shifts by 16. The top 4 bits get lost.

In the end though, it turns out we never use these values for anything (they're all versions specifiers.)

Comment (story645) [inline src/ft2font.h]: Oh,  so they're what the freestype docs call something like FT_MAJOR and FT_MINOR?

Comment (QuLogic) [inline src/ft2font.h]: I don't see such things?

Comment (story645) [inline src/ft2font.h]: https://www.freetype.org/freetype2/docs/reference/ft2-version.html

Comment (QuLogic) [inline src/ft2font.h]: Those are the Freetype version; I'm referring to font file versions.

**Comment by QuLogic** [comment]
Ping @mdboom?

**Comment by QuLogic** [comment]
Ping @mdboom?

File: src/ft2font.h [hunk 2]
```diff
@@ -18,8 +18,8 @@ extern "C" {
 /*
  By definition, FT_FIXED as 2 16bit values stored in a single long.
  */
-#define FIXED_MAJOR(val) (long)((val & 0xffff000) >> 16)
-#define FIXED_MINOR(val) (long)(val & 0xffff)
+#define FIXED_MAJOR(val) (signed short)((val & 0xffff0000) >> 16)
+#define FIXED_MINOR(val) (unsigned short)(val & 0xffff)
```
Comment (anntzer) [inline src/ft2font.h]: Should also cause a type change in the version and fontrevision fields of converting TT_Header no?

**Comment by anntzer** [comment]
I think just a type change is missing?  Otherwise looks fine (but this is the kind of things that decltype in modern C++ (e.g. via pybind11) makes much easier :-)).

File: src/ft2font.h (hunk 2 above)
Comment (QuLogic) [reply src/ft2font.h]: As in [here](https://github.com/matplotlib/matplotlib/pull/7853/commits/ea69e038d026c6272a7af0a309dde5a0606d12ae#diff-22fee04e99c432c6b01d4a2db592cad9)?

**Pull request merged**